File: backend/app/api/v1/items.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db
from app.schemas import ItemCreate, ItemUpdate, ItemResponse, PaginatedResponse
from app.services import ItemService

router = APIRouter()


def get_item_service(db: AsyncSession = Depends(get_db)) -> ItemService:
    return ItemService(db)


@router.get("", response_model=PaginatedResponse[ItemResponse])
async def list_items(
    page: int = Query(1, ge=1),
    size: int = Query(10, ge=1, le=100),
    service: ItemService = Depends(get_item_service),
):
    """获取项目列表"""
    skip = (page - 1) * size
    items, total = await service.get_all(skip=skip, limit=size)
    pages = (total + size - 1) // size if total > 0 else 0

    return PaginatedResponse(
        items=items,
        total=total,
        page=page,
        size=size,
        pages=pages,
    )


@router.get("/{item_id}", response_model=ItemResponse)
async def get_item(
    item_id: int,
    service: ItemService = Depends(get_item_service),
):
    """获取单个项目"""
    item = await service.get_by_id(item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return item


@router.post("", response_model=ItemResponse, status_code=201)
async def create_item(
    data: ItemCreate,
    service: ItemService = Depends(get_item_service),
):
    """创建项目"""
    return await service.create(data)


@router.put("/{item_id}", response_model=ItemResponse)
async def update_item(
    item_id: int,
    data: ItemUpdate,
    service: ItemService = Depends(get_item_service),
):
    """更新项目"""
    item = await service.update(item_id, data)
    if not item:
        raise HTTPException(status_code=404, detail="Item not found")
    return item


@router.delete("/{item_id}", status_code=204)
async def delete_item(
    item_id: int,
    service: ItemService = Depends(get_item_service),
):
    """删除项目"""
    deleted = await service.delete(item_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Item not found")

File: backend/src/routes/items.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { ItemService } from '@/services/item.service';
import {
  createItemSchema,
  updateItemSchema,
  itemParamsSchema,
  listItemsQuerySchema,
} from '@/schemas/item';
import type { Env } from '@/types';

const items = new Hono<{ Bindings: Env }>();

/**
 * GET /api/items
 * List all items with pagination
 */
items.get('/', zValidator('query', listItemsQuerySchema), async (c) => {
  const { page, size } = c.req.valid('query');
  const service = new ItemService(c.get('db'));

  const skip = (page - 1) * size;
  const { items: itemsList, total } = await service.getAll(skip, size);

  const pages = Math.ceil(total / size);

  return c.json({
    items: itemsList,
    total,
    page,
    size,
    pages,
  });
});

/**
 * GET /api/items/:id
 * Get single item by ID
 */
items.get('/:id', zValidator('param', itemParamsSchema), async (c) => {
  const { id } = c.req.valid('param');
  const service = new ItemService(c.get('db'));

  const item = await service.getById(id);

  if (!item) {
    return c.json({ error: 'Item not found' }, 404);
  }

  return c.json(item);
});

/**
 * POST /api/items
 * Create new item
 */
items.post('/', zValidator('json', createItemSchema), async (c) => {
  const data = c.req.valid('json');
  const service = new ItemService(c.get('db'));

  const item = await service.create(data);

  return c.json(item, 201);
});

/**
 * PUT /api/items/:id
 * Update item
 */
items.put(
  '/:id',
  zValidator('param', itemParamsSchema),
  zValidator('json', updateItemSchema),
  async (c) => {
    const { id } = c.req.valid('param');
    const data = c.req.valid('json');
    const service = new ItemService(c.get('db'));

    const item = await service.update(id, data);

    if (!item) {
      return c.json({ error: 'Item not found' }, 404);
    }

    return c.json(item);
  }
);

/**
 * DELETE /api/items/:id
 * Delete item
 */
items.delete('/:id', zValidator('param', itemParamsSchema), async (c) => {
  const { id } = c.req.valid('param');
  const service = new ItemService(c.get('db'));

  const deleted = await service.delete(id);

  if (!deleted) {
    return c.json({ error: 'Item not found' }, 404);
  }

  return c.body(null, 204);
});

export default items;
